### PR TITLE
feat: build images

### PR DESCRIPTION
**Pull request by tedim52**
kurtosis hack week

### CODE REVIEW

File: core/server/api_container/server/startosis_engine/startosis_validator/validated_image.go [hunk 1]
```diff
@@ -1,18 +1,27 @@
...
 func (v *ValidatedImage) GetName() string {
 	return v.name
 }
 
-func (v *ValidatedImage) GetPulledFromRemote() bool {
+func (v *ValidatedImage) IsPulledFromRemote() bool {
 	return v.pulledFromRemote
 }
+
+func (v *ValidatedImage) IsLocallyBuilt() bool {
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/startosis_validator/validated_image.go:25]: change to `isBuiltLocally` to match <verb> <location> of `isPulledFromRemote`

File: container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go [hunk 2]
```diff
@@ -1245,6 +1274,111 @@ func (manager *DockerManager) FetchLatestImage(ctx context.Context, dockerImage
...
+
+	// We generate a new session every time because, per https://github.com/moby/buildkit/issues/1432 ,
+	// sharing sessions is an optimization
+	// We don't bother reusing sessions so that we don't hit bugs
+	buildkitSession, err := session.NewSession(ctx, sessionName, buildkitSessionSharedKey)
+	if err != nil {
+		return stacktrace.Propagate(err, "An error")
+	}
+
+	dialSessionFunc := func(ctx context.Context, proto string, meta map[string][]string) (net.Conn, error) {
```
Comment (tedim52) [inline container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go:1300]: comment what this is doing

Comment (tedim52) [inline container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go:1300]: replace this with the `bkclient.New` function that makes this call under the hood - more explicit that what's actually happening is a client to buildkit is being established

Comment (tedim52) [reply container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go:1300]: https://github.com/moby/moby/blob/311b9ff0aa93aa55880e1e5f8871c4fb69583426/client/buildkit/buildkit.go#L16

File: api/golang/core/lib/enclaves/enclave_context.go [hunk 3]
```diff
@@ -142,12 +155,41 @@ func (enclaveCtx *EnclaveContext) RunStarlarkPackage(
...
+		if composeAbsFilepath == "" {
+			return nil, nil, stacktrace.NewError(
+				"Neither a '%s' file nor one of the default Compose files (%s) was found in the package root; at least one of these is required",
+				kurtosisYamlFilename,
+				strings.Join(supportedDockerComposeYmlFilenames, ", "),
+			)
+		}
+
+		// TODO(kevin): This is a hack, to get around the "only Github URLs" validation!
+		packageName = "github.com/NOTIONAL_USER/USER_UPLOADED_COMPOSE_PACKAGE"
```
Comment (tedim52) [inline api/golang/core/lib/enclaves/enclave_context.go:187]: TODO

File: api/golang/core/lib/enclaves/enclave_context.go [hunk 4]
```diff
@@ -48,6 +48,19 @@ const (
 	dotRelativePathIndicatorString = "."
 )
 
+// TODO(kevin) Remove this once package ID is detected ONLY the APIC side (i.e. the CLI doesn't need to tell the APIC what package ID it's using)
```
Comment (tedim52) [inline api/golang/core/lib/enclaves/enclave_context.go:51]: TODO

File: api/golang/core/lib/enclaves/enclave_context.go [hunk 5]
```diff
@@ -142,12 +155,41 @@ func (enclaveCtx *EnclaveContext) RunStarlarkPackage(
...
+			)
+		}
+
+		// TODO(kevin): This is a hack, to get around the "only Github URLs" validation!
+		packageName = "github.com/NOTIONAL_USER/USER_UPLOADED_COMPOSE_PACKAGE"
+		packageReplaceOptions = make(map[string]string)
 	}
 
-	executeStartosisPackageArgs, err := enclaveCtx.assembleRunStartosisPackageArg(kurtosisYml, runConfig.RelativePathToMainFile, runConfig.MainFunctionName, serializedParams, runConfig.DryRun, runConfig.Parallelism, runConfig.ExperimentalFeatureFlags, runConfig.CloudInstanceId, runConfig.CloudUserId)
+	// TODO(kevin): Remove the package name from the args; we really shouldn't need it!
```
Comment (tedim52) [inline api/golang/core/lib/enclaves/enclave_context.go:191]: TODO (do we need to do this?)

File: api/typescript/src/core/lib/enclaves/enclave_context.ts [hunk 6]
```diff
@@ -151,6 +151,8 @@ export class EnclaveContext {
             return err(new Error(`Unexpected error while getting the Kurtosis yaml file from path '${packageRootPath}'`))
         }
 
+        // TODO TODO TODO TODO put in Compose stuff here!!!
```
Comment (tedim52) [inline api/typescript/src/core/lib/enclaves/enclave_context.ts:154]: TODO

File: cli/cli/helpers/output_printers/kurtosis_instruction_printer.go [hunk 7]
```diff
@@ -107,6 +107,8 @@ func (printer *ExecutionPrinter) PrintKurtosisExecutionResponseLineToStdOut(resp
 		return stacktrace.NewError("Cannot print with a non started printer")
 	}
 
+	logrus.Tracef("Response line from APIC: %+v", responseLine)
```
Comment (tedim52) [inline cli/cli/helpers/output_printers/kurtosis_instruction_printer.go:110]: remove this?

File: core/server/api_container/server/api_container_service.go [hunk 8]
```diff
@@ -79,13 +79,23 @@ var apiContainerPortProtoToPortSpecPortProto = map[kurtosis_core_rpc_api_binding
 	kurtosis_core_rpc_api_bindings.Port_UDP:  port_spec.TransportProtocol_UDP,
 }
 
+var defaultComposeFilenames = []string{
```
Comment (tedim52) [inline core/server/api_container/server/api_container_service.go:82]: is there a place we can centralize this (it's also in the api code)

File: core/server/api_container/server/api_container_service.go [hunk 9]
```diff
@@ -79,13 +79,23 @@ var apiContainerPortProtoToPortSpecPortProto = map[kurtosis_core_rpc_api_binding
...
+}
+
 type ApiContainerService struct {
 	filesArtifactStore *enclave_data_directory.FilesArtifactStore
 
 	serviceNetwork service_network.ServiceNetwork
 
 	startosisRunner *startosis_engine.StartosisRunner
 
+	// TODO rename to packageContentProvider
```
Comment (tedim52) [inline core/server/api_container/server/api_container_service.go:98]: TODO

File: core/server/api_container/server/api_container_service.go [hunk 10]
```diff
@@ -225,59 +237,55 @@ func (apicService *ApiContainerService) InspectFilesArtifactContents(_ context.C
...
 		if err := stream.SendMsg(binding_constructors.NewStarlarkRunResponseLineFromInterpretationError(interpretationError.ToAPIType())); err != nil {
-			return stacktrace.Propagate(err, "Error preparing for package execution and this error could not be sent through the output stream: '%s'", packageId)
+			return stacktrace.Propagate(err, "Error preparing for package execution and this error could not be sent through the output stream: '%s'", packageIdFromArgs)
 		}
 		return nil
 	}
-	packageName := kurtosisYml.GetPackageName()
-	packageReplaceOptions := kurtosisYml.GetPackageReplaceOptions()
-	logrus.Debugf("package replace options received '%+v'", packageReplaceOptions)
+	logrus.Debugf("package replace options received '%+v'", detectedPackageReplaceOptions)
```
Comment (tedim52) [inline core/server/api_container/server/api_container_service.go:275]: TODO: still need this?

File: core/server/api_container/server/api_container_service.go [hunk 11]
```diff
@@ -724,48 +732,97 @@ func (apicService *ApiContainerService) getServiceInfoForIdentifier(ctx context.
 	return serviceInfo, nil
 }
 
+// TODO(kevin) shouldn't this return a normal `error` rather than an `InterpretationError`, because SL interpretation hasn't started yet?
```
Comment (tedim52) [inline core/server/api_container/server/api_container_service.go:735]: TODO: check this

File: core/server/api_container/server/api_container_service.go [hunk 12]
```diff
@@ -724,48 +732,97 @@ func (apicService *ApiContainerService) getServiceInfoForIdentifier(ctx context.
...
+	if transpilationErr != nil {
+		return "", "", "", nil, startosis_errors.WrapWithInterpretationError(transpilationErr, "An error occurred transpiling the Docker Compose package '%v' to Starlark", packageIdFromArgs)
 	}
 
-	return string(mainScriptToExecute), kurtosisYml, nil
+	replacesForComposePackage := map[string]string{}
+
+	return mainScriptToExecute, relativePathToMainFile, packageIdFromArgs, replacesForComposePackage, nil
+
+	// Otherwise, fall back to regular kurtosis.yml parsing
```
Comment (tedim52) [inline core/server/api_container/server/api_container_service.go:825]: TODO: remove?

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+	shouldOverrideComposeYamlKeyProjectName = false
+)
+
+var dockerPortProtosToKurtosisPortProtos = map[string]port_spec.TransportProtocol{
+	"tcp":  port_spec.TransportProtocol_TCP,
+	"udp":  port_spec.TransportProtocol_UDP,
+	"sctp": port_spec.TransportProtocol_SCTP,
+}
+
+// TODO Make this return an interpretation error????
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:48]: TODO

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 14]
```diff
@@ -0,0 +1,496 @@
...
+
+	// Useful for logging, to not leak internals of APIC
+	composeFilename := path.Base(composeRelativeFilepath)
+
+	composeBytes, err := os.ReadFile(composeAbsFilepath)
+	if err != nil {
+		return "", stacktrace.Propagate(err, "An error occurred reading Compose file '%v'", composeFilename)
+	}
+
+	// Use the envvars file next to the Compose if it exists
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:60]: TODO: env vars

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 15]
```diff
@@ -0,0 +1,496 @@
...
+	"strconv"
+	"strings"
+)
+
+const (
+	cpuToMilliCpuConstant = 1024
+	bytesToMegabytes      = 1024 * 1024
+	float64BitWidth       = 64
+
+	// Look for an environment variables file at the package root, and if present use the values found there
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:30]: TODO: handle the case where the .env is not present? Doesn't docker compose have a way to specify the env file to use

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 16]
```diff
@@ -0,0 +1,496 @@
...
+	perServiceLines := map[string][]string{}
+	numTotalServiceLines := 0
+
+	// List of sorted service names that we'll use to make sure we process depends_on correctly
+	sortedServiceNames := []string{}
+
+	for _, serviceConfig := range compose.Services {
+		serviceName := serviceConfig.Name
+
+		// TODO(kevin): handle the dependencyType
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:119]: TODO

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 17]
```diff
@@ -0,0 +1,496 @@
...
+	perServiceDependencies := map[string]map[string]bool{}
+
+	// Mapping of services -> lines of Starlark that they need to run
+	perServiceLines := map[string][]string{}
+	numTotalServiceLines := 0
+
+	// List of sorted service names that we'll use to make sure we process depends_on correctly
+	sortedServiceNames := []string{}
+
+	for _, serviceConfig := range compose.Services {
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:116]: rename `serviceConfig` to `composeServiceConfig` to distinguish from a starlark `ServiceConfig`

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 18]
```diff
@@ -0,0 +1,496 @@
...
+			}
+
+			serviceConfigKwargs = appendKwarg(
+				serviceConfigKwargs,
+				service_config.EnvVarsAttr,
+				enVarsSLDict,
+			)
+		}
+
+		// TODO uncomment
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:226]: TODO

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 19]
```diff
@@ -0,0 +1,496 @@
...
+		}
+
+		argumentValuesSet, interpretationErr := builtin_argument.CreateNewArgumentValuesSet(
+			service_config.ServiceConfigTypeName,
+			service_config.NewServiceConfigType().KurtosisBaseBuiltin.Arguments,
+			[]starlark.Value{},
+			serviceConfigKwargs,
+		)
+		if interpretationErr != nil {
+			// TODO HANDLE THIS! interpretionerror vs go error
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:291]: TODO

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 20]
```diff
@@ -0,0 +1,496 @@
...
+			return "", interpretationErr
+		}
+
+		serviceConfigKurtosisType, interpretationErr := kurtosis_type_constructor.CreateKurtosisStarlarkTypeDefault(service_config.ServiceConfigTypeName, argumentValuesSet)
+		if interpretationErr != nil {
+			// TODO HANDLE THIS! interpretionerror vs go error
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:297]: TODO

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 21]
```diff
@@ -0,0 +1,496 @@
...
+			return "", interpretationErr
+		}
+		serviceConfigStr := serviceConfigKurtosisType.String()
+
+		linesForService := []string{}
+
+		for relativePath, filesArtifactName := range pathsToUpload {
+			// TODO SWITCH FROM HARDCODING THESE TO DYNAMIC CONSTS
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:305]: TODO

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 22]
```diff
@@ -0,0 +1,496 @@
...
+			uploadFilesLine := fmt.Sprintf("plan.upload_files(src = \"%s\", name = \"%s\")", relativePath, filesArtifactName)
+			linesForService = append(linesForService, uploadFilesLine)
+		}
+		// TODO SWITCH FROM HARDCODING THESE TO DYNAMIC CONSTS
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:309]: TODO

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 23]
```diff
@@ -0,0 +1,496 @@
...
+		addServiceLine := fmt.Sprintf("plan.add_service(name = \"%s\", config = %s)", serviceName, serviceConfigStr)
+		linesForService = append(linesForService, addServiceLine)
+
+		perServiceLines[serviceName] = linesForService
+		numTotalServiceLines += len(linesForService)
+	}
+
+	sort.Strings(sortedServiceNames)
+
+	// TODO(kevin) SWITCH THIS TO BE A PROPER DAG!!! This doesn't catch circular depencies
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:319]: TODO

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 24]
```diff
@@ -0,0 +1,496 @@
...
+	)
+
+	argumentValuesSet, interpretationErr := builtin_argument.CreateNewArgumentValuesSet(
+		directory.DirectoryTypeName,
+		directory.NewDirectoryType().KurtosisBaseBuiltin.Arguments,
+		[]starlark.Value{},
+		directoryKwargs,
+	)
+	if interpretationErr != nil {
+		// TODO HANDLE THIS! interpretionerror vs go error
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:485]: TODO

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 25]
```diff
@@ -0,0 +1,496 @@
...
+		return nil, interpretationErr
+	}
+
+	directoryKurtosisType, interpretationErr := kurtosis_type_constructor.CreateKurtosisStarlarkTypeDefault(directory.DirectoryTypeName, argumentValuesSet)
+	if interpretationErr != nil {
+		// TODO FIX THIS! INTERPRETATION ERROR VS GO ERROR
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go:491]: TODO

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+func NewImageBuildSpecType() *kurtosis_type_constructor.KurtosisTypeConstructor {
+	return &kurtosis_type_constructor.KurtosisTypeConstructor{
+		KurtosisBaseBuiltin: &kurtosis_starlark_framework.KurtosisBaseBuiltin{
+			Name: ImageBuildSpecTypeName,
+			Arguments: []*builtin_argument.BuiltinArgument{
+				{
+					Name:              ContextDirAttr,
+					IsOptional:        false,
+					ZeroValueProvider: builtin_argument.ZeroValueProvider[starlark.String],
+					// TODO: add a validator
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go:31]: TODO

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+					Validator: nil,
+				},
+				{
+					Name:              TargetStageAttr,
+					IsOptional:        true,
+					ZeroValueProvider: builtin_argument.ZeroValueProvider[starlark.String],
+					// TODO: add a validator
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go:38]: TODO

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+	kurtosisValueType, err := kurtosis_type_constructor.CreateKurtosisStarlarkTypeDefault(ImageBuildSpecTypeName, arguments)
+	if err != nil {
+		return nil, err
+	}
+	return &ImageBuildSpec{
+		KurtosisValueTypeDefault: kurtosisValueType,
+	}, nil
+}
+
+// ReadyCondition is a starlark.Value that holds all the information needed for ensuring service readiness
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go:57]: rename

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/service_config.go [hunk 29]
```diff
@@ -58,10 +59,13 @@ func NewServiceConfigType() *kurtosis_type_constructor.KurtosisTypeConstructor {
 				{
 					Name:              ImageAttr,
 					IsOptional:        false,
-					ZeroValueProvider: builtin_argument.ZeroValueProvider[starlark.String],
-					Validator: func(value starlark.Value) *startosis_errors.InterpretationError {
-						return builtin_argument.NonEmptyString(value, ImageAttr)
-					},
+					ZeroValueProvider: builtin_argument.ZeroValueProvider[starlark.Value],
+					Validator:         nil,
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/service_config.go:63]: TODO

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/service_config.go [hunk 30]
```diff
@@ -201,6 +205,8 @@ type ServiceConfig struct {
 	*kurtosis_type_constructor.KurtosisValueTypeDefault
 }
 
+// TODO Maybe a CreateServiceConfigUsingGoValues(), as a convenience for transforming Go values to SL values?
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/service_config.go:208]: TODO

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 31]
```diff
@@ -33,17 +33,21 @@ const (
 	defaultDepth = 1
 	// this gets us the entire history - useful for fetching commits on a repo
 	depthAssumingBranchTagsCommitsAreSpecified = 0
-	howImportWorksLink                         = "https://docs.kurtosis.com/explanations/how-do-kurtosis-imports-work"
-	filePathToKurtosisYamlNotFound             = ""
-	replaceCountPackageDirWithGithubConstant   = 1
 
+	// TODO this should be moved to our list of URLs that get unit-tested to ensure they're valid!
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go:37]: TODO

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 32]
```diff
@@ -33,17 +33,21 @@ const (
...
+	howImportWorksLink                       = "https://docs.kurtosis.com/explanations/how-do-kurtosis-imports-work"
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go:38]: TODO

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 33]
```diff
@@ -33,17 +33,21 @@ const (
...
+	filePathToKurtosisYamlNotFound           = ""
+	replaceCountPackageDirWithGithubConstant = 1
+
+	// TODO this should be moved our list of URLs that get unit-tested to ensure they're valid!
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go:42]: TODO

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 34]
```diff
@@ -94,6 +98,7 @@ func (provider *GitPackageContentProvider) GetOnDiskAbsoluteFilePath(absoluteFil
 	if err != nil {
 		return "", startosis_errors.WrapWithInterpretationError(err, "An error occurred parsing Git URL for absolute file locator '%s'", absoluteFileLocator)
 	}
+	// TODO(kevin): Remove this; it will allow users to do `upload_files` on the entire contents of the package (which is very useful for Composes that mount the entire directory like https://github.com/docker/awesome-compose/blob/master/flask-redis/compose.yaml )
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go:101]: TODO

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 35]
```diff
@@ -197,6 +202,8 @@ func (provider *GitPackageContentProvider) StorePackageContents(packageId string
 	if err != nil {
 		return "", startosis_errors.WrapWithInterpretationError(err, "An error occurred while writing contents of '%v' to '%v'", packageId, tempFile.Name())
 	}
+	// TODO(kevin) shouldn't we unarchive to a temp directory and move after, so that if an error occurs we don't pollute the package store with half-extracted
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go:205]: TODO

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_service.go [hunk 36]
```diff
@@ -47,7 +54,7 @@ func NewAddService(serviceNetwork service_network.ServiceNetwork, runtimeValueSt
 					Validator: func(value starlark.Value) *startosis_errors.InterpretationError {
 						// we just try to convert the configs here to validate their shape, to avoid code duplication
 						// with Interpret
-						if _, _, err := validateAndConvertConfigAndReadyCondition(serviceNetwork, value); err != nil {
+						if _, _, err := validateAndConvertConfigAndReadyCondition("", serviceNetwork, value); err != nil {
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_service.go:57]: what is this empty string doing?

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_service_shared.go [hunk 37]
```diff
@@ -65,7 +66,11 @@ func makeAddServiceInterpretationReturnValue(serviceName starlark.String, servic
 	return returnValue, nil
 }
 
-func validateSingleService(validatorEnvironment *startosis_validator.ValidatorEnvironment, serviceName service.ServiceName, serviceConfig *service.ServiceConfig) *startosis_errors.ValidationError {
+func validateSingleService(
+	validatorEnvironment *startosis_validator.ValidatorEnvironment,
+	serviceName service.ServiceName,
+	serviceConfig *service.ServiceConfig,
+	imageBuildSpec *image_build_spec.ImageBuildSpec) *startosis_errors.ValidationError {
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_service_shared.go:73]: why isn't `imageBuildSpec` being passed via the `ServiceConfig`

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_services.go [hunk 38]
```diff
@@ -79,6 +85,9 @@ type AddServicesCapabilities struct {
 
 	readyConditions map[service.ServiceName]*service_config.ReadyCondition
 
+	// TODO: Update add_services to handle building images
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_services.go:88]: TODO

File: core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/builtin_argument/argument_values_set.go [hunk 39]
```diff
@@ -29,6 +30,7 @@ func NewArgumentValuesSet(argumentsDefinition []*BuiltinArgument, values []starl
 	}
 }
 
+// (kevin) Why is this separate from NewArgumentValuesSet?
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/builtin_argument/argument_values_set.go:33]: TODO

**Comment by h4ck3rk3y** [comment]
Closing this as we have the ImageBuildSpec now

**Pull request closed without merging**